File: ApiKit.java
import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.StringReader;
import java.io.UnsupportedEncodingException;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.zip.GZIPInputStream;

import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

import org.apache.commons.codec.binary.Base64;
import org.apache.commons.io.IOUtils;
import org.apache.http.HttpEntity;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.conn.ssl.NoopHostnameVerifier;
import org.apache.http.entity.ContentType;
import org.apache.http.entity.mime.MultipartEntityBuilder;
import org.apache.http.entity.mime.content.StringBody;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.impl.client.LaxRedirectStrategy;
import org.apache.http.util.EntityUtils;
import org.apache.zookeeper.ZooKeeper;
import org.apache.zookeeper.data.Stat;
import org.bouncycastle.openssl.PEMKeyPair;
import org.bouncycastle.openssl.PEMParser;
import org.json.JSONObject;

public class ApiKit {
	public static final int CONNECTTIMEOUT = 7000;
	public static final int READTIMEOUT = 5000;

	/**
	 * @return
	 * @throws Exception
	 */
	public static String getReqUrl(String prefixUrl) throws Exception {
		final String privatekey = "-----BEGIN RSA PRIVATE KEY-----\n"+
			"MIIBOgIBAAJBAKdkYovmgHvxU0Vsz2ft7vVczspAk8jNpMxyQtNJ0ax6BzEZ+svY\n"+
			"R17QjOkGFAM1eZRfJq5rdny+hCYhse8I/9sCAwEAAQJAP0gsGUei+zhYir6ACoJg\n"+
			"/FGBu+R9+kQEMWZg7Q/TPKio1Hlfh50k2PZW5wFBa+PdeyVm9mrUWncx1ZILNw8u\n"+
			"QQIhANKVnsfCsmVrfV+4ojBRHEVauOBoGBdDP+181mjabTltAiEAy34dLD9TN5tC\n"+
			"JlsKnI3rzyQ16rrb0OkxwJqKWPCgWWcCIGxX2kdAXnRbpzd2UMu3D2qHUJL0O2DM\n"+
			"krlm/xEXQBbJAiEAiguc6MZwwrlNr811Lm1MujIbbYij1F5OBRYRonJipSMCIGHa\n"+
			"/YmAFOOo7HfaOui79Qa1ixzSNNaj+9tT3DK1PCf4\n"+
			"-----END RSA PRIVATE KEY-----";
		PEMParser parser = new PEMParser(new StringReader(privatekey));
		PEMKeyPair pari = (PEMKeyPair) parser.readObject();
		parser.close();
		byte[] pk1 = pari.getPrivateKeyInfo().getEncoded();
		String token = "hoops";
		long timestamp = System.currentTimeMillis();

		final SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMddHHmmss");
		String event_time = sdf.format(Calendar.getInstance().getTime());

		byte[] signautreResult = RsaKeyTools.sign(token + timestamp + event_time, pk1);
		String signature = RsaKeyTools.bytes2String(signautreResult);
		return String.format("%s/1.8/%s/%s/%s", prefixUrl, timestamp, event_time, signature);
	}

	/**
	 * 执行网络数据的提交
	 */
	public static JSONObject postJson(String url, byte[] writeBytes) throws Exception {
		return postJson(url, writeBytes, null);
	}

	static class miTM implements javax.net.ssl.TrustManager, javax.net.ssl.X509TrustManager {
		public java.security.cert.X509Certificate[] getAcceptedIssuers() {
			return null;
		}

		public boolean isServerTrusted(java.security.cert.X509Certificate[] certs) {
			return true;
		}

		public boolean isClientTrusted(java.security.cert.X509Certificate[] certs) {
			return true;
		}

		public void checkServerTrusted(java.security.cert.X509Certificate[] certs, String authType)
				throws java.security.cert.CertificateException {
			return;
		}

		public void checkClientTrusted(java.security.cert.X509Certificate[] certs, String authType)
				throws java.security.cert.CertificateException {
			return;
		}
	}

	public static JSONObject postJson(String url, byte[] writeBytes, String cookie) {
		HttpURLConnection con = null;
		JSONObject rsp;
		try {
			if (url.startsWith("https")) {
				javax.net.ssl.HostnameVerifier hv = new javax.net.ssl.HostnameVerifier() {
					public boolean verify(String urlHostName, javax.net.ssl.SSLSession session) {
						return true;
					}
				};
				javax.net.ssl.TrustManager[] trustAllCerts = new javax.net.ssl.TrustManager[1];
				javax.net.ssl.TrustManager tm = new miTM();
				trustAllCerts[0] = tm;
				javax.net.ssl.SSLContext sc = javax.net.ssl.SSLContext.getInstance("SSL");
				sc.init(null, trustAllCerts, null);
				javax.net.ssl.HttpsURLConnection.setDefaultSSLSocketFactory(sc.getSocketFactory());
				HttpsURLConnection.setDefaultHostnameVerifier(hv);
			}
			con = (HttpURLConnection) new URL(url).openConnection();
			con.setConnectTimeout(CONNECTTIMEOUT);
			con.setReadTimeout(READTIMEOUT);
			con.setRequestMethod("POST");
			con.setDoOutput(true);
			con.setRequestProperty("Accept", "application/json");
			con.setRequestProperty("Accept-Encoding", "gzip, deflate");
			con.setRequestProperty("Accept-Language", "zh-cn,zh;q=0.8,en-us;q=0.5,en;q=0.3");
			con.setRequestProperty("Content-Type", "application/json;charset=utf-8");
			con.setRequestProperty("Cache-Control", "max-age=0");
			con.setRequestProperty("Connection", "Keep-Alive");
			if (cookie != null) {
				con.setRequestProperty("Cookie", cookie);
			}
			con.connect();
			if (writeBytes != null) {
				con.getOutputStream().write(writeBytes);
			}
			if (con.getResponseCode() == 200) {
				InputStream is = con.getInputStream();
				boolean gzip = "gzip".equals(con.getHeaderField("Content-Encoding"));
				if (gzip)
					is = new GZIPInputStream(is);
				rsp = new JSONObject(new String(readAsByteArray(is), "UTF-8"));
				if (cookie == null) {
					Map<String, List<String>> fields = con.getHeaderFields();
					Iterator<String> iterator = fields.keySet().iterator();
					while (iterator.hasNext()) {
						String key = iterator.next();
						if (key == null)
							continue;
						List<String> list = fields.get(key);
						StringBuffer val = new StringBuffer();
						for (int i = 0; i < list.size(); i++) {
							if (i > 0) {
								val.append(",");
							}
							val.append(list.get(i));
						}
						rsp.put(key, val.toString());
					}
				}
			} else {
				rsp = new JSONObject();
				rsp.put("errcode", 0);
				byte[] rspbytes = readFullInputStream(con.getInputStream());
				rsp.put("errmsg", String.format("%s(%s)", new String(rspbytes, "UTF-8"), con.getResponseCode()));
				rsp.put("voice", "授权开放");
			}
		} catch (Exception e) {
			e.printStackTrace();
			rsp = new JSONObject();
			rsp.put("errcode", 0);
			rsp.put("errmsg", String.format("处理异常(%s)", e.toString()));
			rsp.put("voice", "授权开放");
		} finally {
			if (con != null) {
				con.disconnect();
			}
		}
		return rsp;
	}

	/**
	 * 读取输入流为二进制数组
	 *
	 * @param servletInputStream
	 * @return
	 */
	public static byte[] readFullInputStream(InputStream is) {
		List<byte[]> all = new ArrayList<byte[]>();
		int read = -1;
		int count = 0;
		byte[] catchs = new byte[1024];
		try {
			while ((read = is.read(catchs)) > -1) {
				count += read;
				byte[] data = new byte[read];
				System.arraycopy(catchs, 0, data, 0, read);
				all.add(data);
			}
			is.close();
		} catch (IOException e) {
			e.printStackTrace();
		}
		byte[] allDatas = new byte[count];
		int copyCounts = 0;
		for (byte[] b : all) {
			System.arraycopy(b, 0, allDatas, copyCounts, b.length);
			copyCounts += b.length;
		}

		return allDatas;
	}

	public static byte[] getByteArrayFromUrl(String urlStr) {
		URL url = null;
		HttpURLConnection connection = null;
		InputStream in = null;
		byte[] content = null;
		try {
			url = new URL(urlStr);
			connection = (HttpURLConnection) url.openConnection();
			connection.setDoInput(true);
			connection.setDoOutput(true);
			connection.setRequestMethod("GET");
			connection.setConnectTimeout(CONNECTTIMEOUT);
			connection.setReadTimeout(READTIMEOUT);
			// 连接
			connection.connect();
			int code = connection.getResponseCode();

			if (code > 400) {
				throw new RuntimeException(String.valueOf(code));
			}
			// 如果返回状态码200
			if (200 == code) {
				// 读取返回数据
				in = connection.getInputStream();
				content = IOUtils.toByteArray(in);
			}

		} catch (MalformedURLException e) {
			e.printStackTrace();
			throw new RuntimeException("不合法的http地址", e);
		} catch (IOException e) {
			e.printStackTrace();
			throw new RuntimeException("读取数据失败", e);
		} finally {
			if (connection != null) {
				connection.disconnect();
			}
			if (in != null) {
				try {
					in.close();
				} catch (IOException e) {
					e.printStackTrace();
					throw new RuntimeException("关闭流失败", e);
				}
			}
		}
		return content;
	}

	/**
	 * 得到
	 *
	 * @param path
	 * @param decrypte
	 * @return
	 */
	public static JSONObject getJSONObject(ZooKeeper zookeeper, String path) {
		try {
			byte[] payload = null;
			if (zookeeper != null) {
				Stat stat = zookeeper.exists(path, false);
				payload = zookeeper.getData(path, false, stat);
			} else {
				File file = new File(String.format("../data%s/json.txt", path));
				if (file.exists()) {
					payload = readAsByteArray(file);
				}
			}
			if (payload != null) {
				return new JSONObject(new String(payload, "UTF-8"));
			}
		} catch (Exception e) {
		}
		return null;
	}

	/**
	 * 上传二进制照片
	 * 
	 * @title
	 * @methodName
	 * @author wangyi
	 * @description
	 * @param string 通过getReqUrl升成的URL
	 * @param file  上传文件的数据
	 * @param filename 上传文件名（随便）
	 * @param hashMap
	 * @return
	 * @throws Exception
	 */
	public static JSONObject postImage(String serverUrl, byte[] file, String filename, Map<String, Object> params)
			throws Exception {
		String respStr = null;
		SSLContext sslContext = SSLContext.getInstance("TLS");
		sslContext.init(null, new TrustManager[] { new X509TrustManager() {
			public void checkClientTrusted(X509Certificate[] arg0, String arg1) throws CertificateException {
			}

			public void checkServerTrusted(X509Certificate[] arg0, String arg1) throws CertificateException {

			}

			public X509Certificate[] getAcceptedIssuers() {
				return new X509Certificate[0];
			}
		} }, new java.security.SecureRandom());
		CloseableHttpClient httpclient = HttpClients.custom().setRedirectStrategy(new LaxRedirectStrategy())
				.setSSLContext(sslContext).setSSLHostnameVerifier(new NoopHostnameVerifier()).build();
		try {

			HttpPost httppost = new HttpPost(serverUrl);
			MultipartEntityBuilder multipartEntityBuilder = MultipartEntityBuilder.create();
			multipartEntityBuilder.addBinaryBody("file", new ByteArrayInputStream(file),
					ContentType.create("multipart/form-data"), filename);
			// 设置上传的其他参数
			setUploadParams(multipartEntityBuilder, params);
			HttpEntity reqEntity = multipartEntityBuilder.build();
			httppost.setEntity(reqEntity);
			CloseableHttpResponse response = httpclient.execute(httppost);
			try {
				HttpEntity resEntity = response.getEntity();
				respStr = getRespString(resEntity);
				EntityUtils.consume(resEntity);
			} finally {
				response.close();
			}
			return new JSONObject(respStr);
		} catch (Exception e) {
			throw e;
		} finally {
			try {
				httpclient.close();
			} catch (Exception e) {
			}
		}
	}

	/**
	 * 设置上传文件时所附带的其他参数
	 * 
	 * @param multipartEntityBuilder
	 * @param params
	 * @throws UnsupportedEncodingException
	 */
	public static void setUploadParams(MultipartEntityBuilder multipartEntityBuilder, Map<String, Object> params)
			throws UnsupportedEncodingException {
		if (params != null && params.size() > 0) {
			Set<String> keys = params.keySet();
			for (String key : keys) {
				multipartEntityBuilder.addPart(key, new StringBody(params.get(key).toString(), ContentType.TEXT_PLAIN));
			}
		}
	}
	/**
	 * 将返回结果转化为String
	 * 
	 * @param entity
	 * @return
	 * @throws Exception
	 */
	public static String getRespString(HttpEntity entity) throws Exception {
		if (entity == null) {
			return null;
		}
		InputStream is = entity.getContent();
		StringBuffer strBuf = new StringBuffer();
		byte[] buffer = new byte[4096];
		int r = 0;
		while ((r = is.read(buffer)) > 0) {
			strBuf.append(new String(buffer, 0, r, "UTF-8"));
		}
		return strBuf.toString();
	}

	/**
	 * 还原Unicode编码的字符串
	 * 
	 * @param src
	 * @return String
	 * @author Focus
	 */
	public static String unicode2Chr(String src) {
		if (src == null || src.isEmpty()) {
			return src;
		}
		String tempStr = "";
		String returnStr = "";

		// 将编码过的字符串进行重排
		for (int i = 0; i < src.length() / 4; i++) {
			if (0 == i) {
				tempStr = src.substring(4 * i + 2, 4 * i + 4);
				tempStr += src.substring(4 * i, 4 * i + 2);
			} else {
				tempStr += src.substring(4 * i + 2, 4 * i + 4);
				tempStr += src.substring(4 * i, 4 * i + 2);
			}
		}

		byte[] b = new byte[tempStr.length() / 2];

		try {
			// 将重排过的字符串放入byte数组，用于进行转码
			for (int j = 0; j < tempStr.length() / 2; j++) {
				String subStr = tempStr.substring(j * 2, j * 2 + 2);
				int b1 = Integer.decode("0x" + subStr).intValue();
				b[j] = (byte) b1;
			}
			// 转码
			returnStr = new String(b, "utf-16");
		} catch (UnsupportedEncodingException ex) {
			ex.printStackTrace();
		} catch (Exception e) {
			return src;
		}
		return returnStr;
	}

	/**
	 * 字符串准成unicode
	 * 
	 * @param src
	 * @return
	 */
	public static String chr2Unicode(String src) {
		StringBuffer s = new StringBuffer();
		if (src != null && !"".equals(src)) {
			String hex = "";

			for (int i = 0; i < src.length(); i++) {
				hex = Integer.toHexString((int) src.charAt(i));
				if (hex.length() < 4) {
					while (hex.length() < 4) {
						hex = "0".concat(hex);
					}
				}
				hex = hex.substring(2, 4).concat(hex.substring(0, 2));
				s.append(hex.toUpperCase());
			}
		}
		return s.toString();
	}

	public static final byte[] readAsByteArray(File file) {
		if (file != null && file.exists()) {
			try {
				FileInputStream fis = new FileInputStream(file);
				byte[] buffer = new byte[fis.available()];
				fis.read(buffer);
				fis.close();
				return buffer;
			} catch (Exception e) {
				return new byte[0];
			}
		} else {
			return new byte[0];
		}
	}

	public static final byte[] readAsByteArray(InputStream is) throws Exception {
		List<byte[]> all = new ArrayList<byte[]>();
		int read = -1;
		int count = 0;
		byte[] catchs = new byte[65536];
		try {
			while ((read = is.read(catchs)) > -1) {
				count += read;
				byte[] data = new byte[read];
				System.arraycopy(catchs, 0, data, 0, read);
				all.add(data);
			}
		} catch (IOException e) {
		} finally {
			is.close();
		}

		byte[] allDatas = new byte[count];
		int copyCounts = 0;
		for (byte[] b : all) {
			System.arraycopy(b, 0, allDatas, copyCounts, b.length);
			copyCounts += b.length;
		}

		return allDatas;
	}
}

File: app/src/main/kotlin/com/smartbasketball/app/util/ApiKit.kt
package com.smartbasketball.app.util

import java.io.ByteArrayInputStream
import java.io.File
import java.io.FileInputStream
import java.io.InputStream
import java.io.StringReader
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.zip.GZIPInputStream
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager
import org.bouncycastle.openssl.PEMKeyPair
import org.bouncycastle.openssl.PEMParser
import org.json.JSONObject

object ApiKit {
    const val CONNECTTIMEOUT = 7000
    const val READTIMEOUT = 5000
    
    private val privateKey = """
-----BEGIN RSA PRIVATE KEY-----
MIIBOgIBAAJBAKdkYovmgHvxU0Vsz2ft7vVczspAk8jNpMxyQtNJ0ax6BzEZ+svY
R17QjOkGFAM1eZRfJq5rdny+hCYhse8I/9sCAwEAAQJAP0gsGUei+zhYir6ACoJg
/FGBu+R9+kQEMWZg7Q/TPKio1Hlfh50k2PZW5wFBa+PdeyVm9mrUWncx1ZILNw8u
QQIhANKVnsfCsmVrfV+4ojBRHEVauOBoGBdDP+181mjabTltAiEAy34dLD9TN5tC
JlsKnI3rzyQ16rrb0OkxwJqKWPCgWWcCIGxX2kdAXnRbpzd2UMu3D2qHUJL0O2DM
krlm/xEXQBbJAiEAiguc6MZwwrlNr811Lm1MujIbbYij1F5OBRYRonJipSMCIGHa
/YmAFOOo7HfaOui79Qa1ixzSNNaj+9tT3DK1PCf4
-----END RSA PRIVATE KEY-----
""".trimIndent()

    @Throws(Exception::class)
    fun getReqUrl(prefixUrl: String): String {
        val parser = PEMParser(StringReader(privateKey))
        val pari = parser.readObject() as PEMKeyPair
        parser.close()
        val pk1 = pari.privateKeyInfo.encoded
        val token = "hoops"
        val timestamp = System.currentTimeMillis()
        val sdf = SimpleDateFormat("yyyyMMddHHmmss")
        val eventTime = sdf.format(Calendar.getInstance().time)

        val signautreResult = RsaKeyTools.sign(token + timestamp + eventTime, pk1)
        val signature = RsaKeyTools.bytes2String(signautreResult)
        val fullUrl = String.format("%s/1.8/%s/%s/%s", prefixUrl, timestamp, eventTime, signature)
        
        AppLogger.d("========== 生成请求URL ==========")
        AppLogger.d("接口URL: $fullUrl")
        
        return fullUrl
    }

    @Throws(Exception::class)
    fun postJson(url: String, writeBytes: ByteArray?): JSONObject {
        return postJson(url, writeBytes, null)
    }

    fun postJson(url: String, writeBytes: ByteArray?, cookie: String?): JSONObject {
        var con: HttpURLConnection? = null
        var rsp: JSONObject = JSONObject()
        
        AppLogger.d("========== 发送POST请求 ==========")
        AppLogger.d("请求URL: $url")
        if (writeBytes != null) {
            val requestBody = String(writeBytes, Charsets.UTF_8)
            AppLogger.d("请求Body: $requestBody")
        }
        
        try {
            if (url.startsWith("https")) {
                val hv = javax.net.ssl.HostnameVerifier { _, _ -> true }
                val trustAllCerts = arrayOf<TrustManager>(object : X509TrustManager {
                    override fun getAcceptedIssuers(): Array<java.security.cert.X509Certificate>? = null
                    override fun checkClientTrusted(certs: Array<java.security.cert.X509Certificate>?, authType: String?) {}
                    override fun checkServerTrusted(certs: Array<java.security.cert.X509Certificate>?, authType: String?) {}
                })
                val sc = SSLContext.getInstance("SSL")
                sc.init(null, trustAllCerts, null)
                HttpsURLConnection.setDefaultSSLSocketFactory(sc.socketFactory)
                HttpsURLConnection.setDefaultHostnameVerifier(hv)
            }
            con = URL(url).openConnection() as HttpURLConnection
            con.connectTimeout = CONNECTTIMEOUT
            con.readTimeout = READTIMEOUT
            con.requestMethod = "POST"
            con.doOutput = true
            con.setRequestProperty("Accept", "application/json")
            con.setRequestProperty("Accept-Encoding", "gzip, deflate")
            con.setRequestProperty("Accept-Language", "zh-cn,zh;q=0.8,en-us;q=0.5,en;q=0.3")
            con.setRequestProperty("Content-Type", "application/json;charset=utf-8")
            con.setRequestProperty("Cache-Control", "max-age=0")
            con.setRequestProperty("Connection", "Keep-Alive")
            if (cookie != null) {
                con.setRequestProperty("Cookie", cookie)
            }
            con.connect()
            if (writeBytes != null) {
                con.outputStream.write(writeBytes)
            }
            if (con.responseCode == 200) {
                var inputStream: InputStream = con.inputStream
                val gzip = "gzip" == con.getHeaderField("Content-Encoding")
                if (gzip) {
                    inputStream = GZIPInputStream(inputStream)
                }
                rsp = JSONObject(String(readAsByteArray(inputStream), Charsets.UTF_8))
            } else {
                rsp = JSONObject()
                rsp.put("errcode", 0)
                rsp.put("errmsg", "${String(readAsByteArray(con.errorStream), Charsets.UTF_8)}(${con.responseCode})")
                rsp.put("voice", "授权开放")
            }
        } catch (e: Exception) {
            e.printStackTrace()
            rsp = JSONObject()
            rsp.put("errcode", 0)
            rsp.put("errmsg", "处理异常(${e.message})")
            rsp.put("voice", "授权开放")
        } finally {
            con?.disconnect()
        }
        return rsp
    }

    fun getByteArrayFromUrl(urlStr: String): ByteArray? {
        return try {
            val url = URL(urlStr)
            val connection = url.openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            connection.connectTimeout = CONNECTTIMEOUT
            connection.readTimeout = READTIMEOUT
            connection.connect()
            val code = connection.responseCode
            if (code == 200) {
                connection.inputStream.use { it.readBytes() }
            } else {
                null
            }
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    fun readAsByteArray(file: File): ByteArray {
        return if (file.exists()) {
            FileInputStream(file).use { it.readBytes() }
        } else {
            ByteArray(0)
        }
    }

    @Throws(Exception::class)
    fun readAsByteArray(inputStream: InputStream): ByteArray {
        return inputStream.use { it.readBytes() }
    }

    @Throws(Exception::class)
    fun postImage(serverUrl: String, file: ByteArray, filename: String, params: Map<String, Any>): JSONObject {
        AppLogger.d("========== postImage上传图片 ==========")
        AppLogger.d("URL: $serverUrl, 文件大小: ${file.size}, 文件名: $filename")
        
        val boundary = "----WebKitFormBoundary" + System.currentTimeMillis()
        val lineEnd = "\r\n"
        
        try {
            if (serverUrl.startsWith("https")) {
                val hv = javax.net.ssl.HostnameVerifier { _, _ -> true }
                val trustAllCerts = arrayOf<TrustManager>(object : X509TrustManager {
                    override fun getAcceptedIssuers(): Array<java.security.cert.X509Certificate>? = null
                    override fun checkClientTrusted(certs: Array<java.security.cert.X509Certificate>?, authType: String?) {}
                    override fun checkServerTrusted(certs: Array<java.security.cert.X509Certificate>?, authType: String?) {}
                })
                val sc = SSLContext.getInstance("SSL")
                sc.init(null, trustAllCerts, null)
                HttpsURLConnection.setDefaultSSLSocketFactory(sc.socketFactory)
                HttpsURLConnection.setDefaultHostnameVerifier(hv)
            }
            
            val connection = URL(serverUrl).openConnection() as HttpsURLConnection
            connection.connectTimeout = CONNECTTIMEOUT
            connection.readTimeout = READTIMEOUT
            connection.requestMethod = "POST"
            connection.doOutput = true
            connection.setRequestProperty("Connection", "Keep-Alive")
            connection.setRequestProperty("Content-Type", "multipart/form-data; boundary=$boundary")
            
            val outputStream = java.io.DataOutputStream(connection.outputStream)
            
            // 添加params参数
            for ((key, value) in params) {
                outputStream.writeBytes("--$boundary$lineEnd")
                outputStream.writeBytes("Content-Disposition: form-data; name=\"$key\"$lineEnd$lineEnd")
                outputStream.writeBytes("$value$lineEnd")
            }
            
            // 添加文件
            outputStream.writeBytes("--$boundary$lineEnd")
            outputStream.writeBytes("Content-Disposition: form-data; name=\"file\"; filename=\"$filename\"$lineEnd")
            outputStream.writeBytes("Content-Type: image/jpeg$lineEnd$lineEnd")
            outputStream.write(file)
            outputStream.writeBytes(lineEnd)
            
            // 结束
            outputStream.writeBytes("--$boundary--$lineEnd")
            outputStream.flush()
            outputStream.close()
            
            val responseCode = connection.responseCode
            val responseStr = if (responseCode == 200) {
                String(connection.inputStream.readBytes(), Charsets.UTF_8)
            } else {
                String(connection.errorStream?.readBytes() ?: ByteArray(0), Charsets.UTF_8)
            }
            
            AppLogger.d("postImage响应: $responseStr")
            connection.disconnect()
            
            return JSONObject(responseStr)
        } catch (e: Exception) {
            AppLogger.e("postImage异常: ${e.message}")
            val rsp = JSONObject()
            rsp.put("errcode", -1)
            rsp.put("errmsg", "上传异常(${e.message})")
            return rsp
        }
    }
}
